site/content/docs/user/exposing-applications.md

```diff
@@ -0,0 +1,242 @@
+
+### ClusterIP
+
+> Exposes the service on a cluster-internal IP. Choosing this value makes the service only reachable from within the cluster. This is the default ServiceType
+
+```sh
+kubectl expose deployment hello-world --port=80 --target-port=8080 --name=hello-world
+
+service/hello-world exposed
+```
+
+```
+kubectl describe services hello-world
+
+Name:              hello-world
+Namespace:         default
+Labels:            run=load-balancer-example
+Annotations:       <none>
+Selector:          run=load-balancer-example
+Type:              ClusterIP
+IP:                10.106.26.187
+Port:              <unset>  80/TCP
+TargetPort:        8080/TCP
+Endpoints:         10.36.0.1:8080,10.44.0.1:8080
+Session Affinity:  None
+Events:            <none>
+```
+
+### NodePort
+
+> NodePort, as the name implies, opens a specific port on all the Nodes (the VMs), and any traffic that is sent to this port is forwarded to the service.
+
+
+```
+kubectl expose deployment hello-world --type=NodePort --name=hello-world-nodeport
+
+service/hello-world-nodeport exposed
+```
+
+```
+kubectl describe services hello-world-nodeport
+Name:                     hello-world-nodeport
+Namespace:                default
+Labels:                   run=load-balancer-example
+Annotations:              <none>
+Selector:                 run=load-balancer-example
+Type:                     NodePort
+IP:                       10.96.59.230
+Port:                     <unset>  8080/TCP
+TargetPort:               8080/TCP
+NodePort:                 <unset>  30039/TCP
+Endpoints:                10.36.0.1:8080,10.44.0.1:8080
+Session Affinity:         None
+External Traffic Policy:  Cluster
+Events:                   <none>
+```
+
+
+### Nginx ingress controller
+
+> Ingress is actually NOT a type of service. Instead, it sits in front of multiple services and act as a “smart router” or entrypoint into your cluster.
+
+TODO
+
+## Accesing the application
+
+For some of this methods we'll need the cluster IP addresses
+
+```
+kubectl get nodes -o wide
+
+NAME                 STATUS   ROLES    AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                                  KERNEL-VERSION     CONTAINER-RUNTIME
+kind-control-plane   Ready    master   7m1s    v1.14.1   172.17.0.4    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+kind-worker          Ready    <none>   5m58s   v1.14.1   172.17.0.2    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+kind-worker2         Ready    <none>   5m59s   v1.14.1   172.17.0.3    <none>        Ubuntu Disco Dingo (development branch)   4.9.125-linuxkit   containerd://1.2.6-0ubuntu1
+```
+
+### Kubectl port-forward
+
+This is the most simple one and well documented
+
+https://kubernetes.io/docs/tasks/access-application-cluster/port-forward-access-application-cluster/
+
+### Port forwarding with socat
```

Comment:
This can be socat or our nginx load balancer, the trick here is to keep adding ports to the same container, i.e.:

Expose localhost 8080 on nodeA:9090
docker run -p 8080 node_proxy "command to proxy to nodeA9090"

if we want to add a new port we recreate the container with the new data
docker run -p 8080 -p <new_port> node_proxy "command to proxy to nodeA9090" "command to proxy <newport>"